strategies/categorized/src/test/java/org/arquillian/smart/testing/strategies/categorized/CategorizedTestsDetectorTest.java

```diff
@@ -213,6 +208,94 @@ public void should_return_classes_and_methods_without_first_category_when_revers
+
+        assertThat(testSelection).containsTestClassSelectionsExactlyInAnyOrder(testSelectionWithMethods);
+    }
+
+    @Test
+    public void should_return_methods_without_excluded_tag() {
+        //given
+        categorizedConfig.setCategories(new String[] {"first"});
+        categorizedConfig.setExcludedCategories(new String[] {"second","third"});
+
+        //when
+        Collection<TestSelection> testSelection =
+            new CategorizedTestsDetector(config).selectTestsFromClasses(Collections.singleton(WithTaggedMethodsClass.class));
+
+        //then
+        TestSelection testSelectionWithMethods = getTestSelectionWithMethods(WithTaggedMethodsClass.class, "testWithFirstTag");
+
+        assertThat(testSelection).containsTestClassSelectionsExactlyInAnyOrder(testSelectionWithMethods);
+    }
+
+    @Test
+    public void should_return_class_when_no_methods_are_matched_to_be_excluded() {
+        //given
+        categorizedConfig.setCategories(new String[] {FirstCategory.class.getSimpleName()});
+        categorizedConfig.setExcludedCategories(new String[] {Excluded.class.getSimpleName()});
+
+        //when
+        Collection<TestSelection> testSelection =
+            new CategorizedTestsDetector(config).selectTestsFromClasses(Collections.singleton(FirstCategorizedClass.class));
+
+        assertThat(testSelection).containsTestClassSelectionsExactlyInAnyOrder(getTestSelection(FirstCategorizedClass.class));
+    }
+
+    @Test
+    public void should_not_exclude_test_class_when_excluding_single_category_of_many() {
+        //given
+        categorizedConfig.setCategories(new String[] {FirstCategory.class.getSimpleName()});
+        categorizedConfig.setExcludedCategories(new String[] {SecondCategory.class.getSimpleName()});
+
+        //when
+        Collection<TestSelection> testSelection =
+            new CategorizedTestsDetector(config).selectTestsFromClasses(Collections.singleton(FirstAndSecondCategorizedClass.class));
+
+        assertThat(testSelection).containsTestClassSelectionsExactlyInAnyOrder(getTestSelection(FirstAndSecondCategorizedClass.class));
+    }
+
+    @Test
+    public void should_exclude_test_class_when_excluding_multiple_categories_and_match_all_is_set() {
+        //given
+        categorizedConfig.setCategories(new String[] {FirstCategory.class.getSimpleName()});
+        categorizedConfig.setExcludedCategories(new String[] {SecondCategory.class.getSimpleName(), FirstCategory.class.getSimpleName()});
+        categorizedConfig.setMatchAll(true);
+
+        //when
+        Collection<TestSelection> testSelection =
+            new CategorizedTestsDetector(config).selectTestsFromClasses(Collections.singleton(FirstAndSecondCategorizedClass.class));
+
+        assertThat(testSelection).containsTestClassSelectionsExactlyInAnyOrder(getTestSelection(FirstAndSecondCategorizedClass.class));
+    }
+
+    @Test
+    public void should_exclude_methods_when_matched_and_class_is_not_annotated() {
```

Comment:
this test method doesn't test the exclude logic - the resulting set of methods will contain always only the method `testWithFirstCategory` because only this one has the `FirstCategory` specified in included categories.